scheds/rust/scx_rusty/src/bpf/main.bpf.c

```diff
@@ -657,6 +712,39 @@ s32 BPF_STRUCT_OPS(rusty_select_cpu, struct task_struct *p, s32 prev_cpu,
 		}
 	}
 
+	/*
+	 * This checks if there are any CPUs in the task's domain that can be
+	 * preempted: e.g. if the task's vtime is less the highest vtime (vtime_max)
+	 * in the domain (minus a small buffer: slice_ns). If there is, then set
+	 * the vtime_max for the domain to the next highest vtime running on the
+	 * domain's CPUs. This is to increase responsiveness for higher priority
+	 * tasks. The task might also be forked/execd so check for wake up only.
+	 */
+	if (!no_preemption && !fifo_sched && (wake_flags & SCX_WAKE_TTWU)) {
+		u32 dom_id = taskc->dom_id;
+		struct dom_ctx *domc;
+
+		if (!(domc = bpf_map_lookup_elem(&dom_data, &dom_id))) {
+			scx_bpf_error("Failed to lookup dom[%u]", dom_id);
+			goto enoent;
+		}
+
```

Comment:
I tried that at first but the verifier wasn't happy with me. I don't think it's that smart

Reply:
hmmm, that should work. I've done that in a number of other places. this really doesn't work?

```
if (!preempted_cpumask) {
    scx_bpf_error("No cpumask found");
    goto enoent;
}
```

could you please share the verifier output?

Reply:
So I believe it's because the use of the preempted_cpumask must be part of the next instruction or else it can't guarantee that it hasn't been modified in another function call e.g. this works:

```
if (!preempted_cpumask) {
  scx_bpf_error("No preempted_cpumask found");
   goto enoent;
}

if (!bpf_cpumask_test_cpu(domc->vtime_max_cpu, preempted_cpumask)
&& bpf_cpumask_test_cpu(domc->vtime_max_cpu, p->cpus_ptr)
...
```

But this does not:
```
```
if (!preempted_cpumask) {
  scx_bpf_error("No preempted_cpumask found");
   goto enoent;
}

if (bpf_cpumask_test_cpu(domc->vtime_max_cpu, p->cpus_ptr)
&& !bpf_cpumask_test_cpu(domc->vtime_max_cpu, preempted_cpumask)
...
```

Here is the end of the verifier log:
```
; && !bpf_cpumask_test_cpu(domc->vtime_max_cpu, preempted_cpumask)
515: (16) if w0 == 0x0 goto pc+40     ; R0=scalar(smin=smin32=0,smax=umax=smax32=umax32=1,var_off=(0x0; 0x1)) refs=1
; && !bpf_cpumask_test_cpu(domc->vtime_max_cpu, preempted_cpumask)
516: (18) r1 = 0xffff88810258e328     ; R1_w=map_value(off=24,ks=4,vs=32,imm=0) refs=1
518: (79) r2 = *(u64 *)(r1 +0)        ; R1_w=map_value(off=24,ks=4,vs=32,imm=0) R2_w=rcu_ptr_or_null_bpf_cpumask(id=29,off=0,imm=0) refs=1
; && !bpf_cpumask_test_cpu(domc->vtime_max_cpu, preempted_cpumask)
519: (61) r1 = *(u32 *)(r9 +16)       ; R1_w=scalar(smin=0,smax=umax=4294967295,var_off=(0x0; 0xffffffff)) R9=map_value(off=0,ks=4,vs=4048,imm=0) refs=1
; && !bpf_cpumask_test_cpu(domc->vtime_max_cpu, preempted_cpumask)
520: (85) call bpf_cpumask_test_cpu#132690
Possibly NULL pointer passed to trusted arg1
processed 401 insns (limit 1000000) max_states_per_insn 2 total_states 42 peak_states 42 mark_read 18
-- END PROG LOAD LOG --
```

Reply:
ah, yes, you have to copy it into a local variable. but that's fine, because you know the global variable isn't actually going to change to another cpumask. it's just a way to get the verifier to not worry about the cpumask being possibly null due to being RCU protected. so something like this should work: https://github.com/sched-ext/scx/blob/main/scheds/c/scx_nest.bpf.c#L436-L441

Reply:
Ah right right. Sometimes my brain doesn't handle pointer copying correctly 😕